File: upes DSA Assignments/Stack/implementation/implementationUsingArray.py

class StackUsingArray:
    
    def __init__(self,size) -> None:
        self.stack = []
        self.size = size
        
    def isFull(self):
        return self.size == len(self.stack) 
    
    def isEmpty(self):
        return len(self.stack) == 0   
    
    def push(self, item):
        if self.isFull():
            print("Stack is full, cannot insert new element")
            
        else:
            self.stack.append(item)
            print(f"pushed {item} to the stack")
            
    
    def pop(self):
        if self.isEmpty():
            print("Stack is empty, cannot delete an element")
            return
        
        else:
            self.stack.pop();
            print(f"{self.stack[-1]} is popped from stack")
            self.size -= 1                
            
    def peek(self):
        
        if self.isEmpty():
            print("Stack is empty")
            return
        
        else:
            print("Top element is: ",self.stack[-1])
            return
        
    def len(self):
        print("length of stack is: ", self.size)    
    
    def display(self):
         if self.isEmpty():
             print("stack is empty")
             
         else:
             print(self.stack[::-1])
  
  
             
s = StackUsingArray(4) # creating the stack of size 4
s.push(10)
s.push(20)
s.push(30)
s.push(40)
s.push(50) # output: Stack is full, cannot insert new element
s.len() # output: length of stack is: 4 
s.display() # output: [40,30,20,10]

s.peek() # output: 40

s.pop() # output: 30 is popped from stack
s.len() # output: length of stack is 3
s.display() #output: [30,20,10]
             
                    

File: upes DSA Assignments/Stack/implementationUsingArray.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 5 // Define the maximum size of the stack

int stack[MAX]; // Array to hold stack elements
int top = -1; // Initialize top of stack to -1 (indicating stack is empty)

// Function to check if the stack is empty
int isEmpty() {
    return top == -1;
}

// Function to check if the stack is full
int isFull() {
     return top == MAX - 1;
}

// Function to push an element onto the stack
void push(int data)
{
     if (isFull())
     {
          printf("Stack overflow\n");
          return;
     }
     else
     {
          top++;
          stack[top] = data;
          printf("pushed %d in the stack\n", data);
     }
}

// Function to pop an element from the stack
int pop()
{
     if (isEmpty())
     {
          printf("stack underflow\n");
          return -1;
     }
     else
     {
          int popped = stack[top];
          top--;
          return popped;
     }
}

// Function to peek the top element of the stack
int peek()
{
     if (isEmpty())
     {
          printf("stack underflow\n");
          return -1;
     }
     else
     {
          return stack[top];
     }
}

// Function to display all elements in the stack
void display()
{
     if (isEmpty())
     {
          printf("stack is empty\n");
          return;
     }
     else
     {
          printf("stack elements:\n");
          for (int i = top; i >= 0; i--)
          {
                printf("%d ", stack[i]);
          }
          printf("\n");
     }
}

int main()
{
     printf("Stack implementation using array\n");

     int choice, value;

     while (1)
     {
         
          printf("1 for push\n");
          printf("2 for pop\n");
          printf("3 for peek\n");
          printf("4 display\n");
          printf("5 exit\n");
          printf("enter the choice: ");
          scanf("%d", &choice);

          switch (choice)
          {
          case 1:
                printf("enter the element to push into stack\n");
                scanf("%d", &value);
                push(value);
                break;

          case 2:
                printf("\npopped value from stack is %d\n", pop());
                break;

          case 3:
                printf("\ntop element from stack %d\n", peek());
                break;

          case 4:
                display();
                break;

          case 5:
                exit(0);
                break;

          default:
                printf("Invalid choice! please try again\n");
          }
     }

     return 0;
}